File: app/src/main/java/br/com/jonathanzanella/myexpenses/database/Table.java
package br.com.jonathanzanella.myexpenses.database;

import android.content.ContentValues;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;

import br.com.jonathanzanella.myexpenses.sync.UnsyncModel;

/**
 * Created by jzanella on 11/1/16.
 */

public interface Table <T extends UnsyncModel> {
	void onCreate(SQLiteDatabase db);
	void onUpgrade(SQLiteDatabase sqLiteDatabase, int oldVersion, int newVersion);

	String getName();
	ContentValues fillContentValues(T data);
	T fill(Cursor c);
	String [] getProjection();
}

File: app/src/main/java/br/com/jonathanzanella/myexpenses/database/Table.kt
package br.com.jonathanzanella.myexpenses.database

import android.content.ContentValues
import android.database.Cursor
import android.database.sqlite.SQLiteDatabase

interface Table<T> {
    fun onCreate(db: SQLiteDatabase)
    fun onUpgrade(sqLiteDatabase: SQLiteDatabase, oldVersion: Int, newVersion: Int)
    fun onDrop(db: SQLiteDatabase)

    fun fillContentValues(data: T): ContentValues
    fun fill(c: Cursor): T

    val name: String
    val projection: Array<String>
}